### PR TITLE
mon/auth: trigger_propose not active cause mon leader update rotate r…

### PR DESCRIPTION
**Pull request by swg1991**
When Paxos in the update state, mon leader need update rotate secret just right. It cause trigger_propose not active and return false. update rotate secret is not timely. For this case，just retry propose . retry three times at most,if retry three times still failed just give up retransmission。

Fixes: https://tracker.ceph.com/issues/50775

Signed-off-by: wenge.song <wenge.song@lenovonetapp.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by swg1991** [comment]
@idryomov ping?

File: src/auth/cephx/CephxKeyServer.cc [hunk 1]
```diff
@@ -151,16 +151,16 @@ int KeyServer::start_server()
...
 {
   ldout(cct, 10) << "_check_rotating_secrets" << dendl;
 
   int added = 0;
-  added += _rotate_secret(CEPH_ENTITY_TYPE_AUTH);
-  added += _rotate_secret(CEPH_ENTITY_TYPE_MON);
-  added += _rotate_secret(CEPH_ENTITY_TYPE_OSD);
-  added += _rotate_secret(CEPH_ENTITY_TYPE_MDS);
-  added += _rotate_secret(CEPH_ENTITY_TYPE_MGR);
+  added += _rotate_secret(CEPH_ENTITY_TYPE_AUTH,auth_done);
```
Comment (idryomov) [inline src/auth/cephx/CephxKeyServer.cc:159]:
```suggestion
  added += _rotate_secret(CEPH_ENTITY_TYPE_AUTH, auth_done);
```

File: src/auth/cephx/CephxKeyServer.cc [hunk 2]
```diff
@@ -210,9 +210,11 @@ int KeyServer::_rotate_secret(uint32_t service_id)
     uint64_t secret_id = r.add(ek);
     ldout(cct, 10) << "_rotate_secret adding " << ceph_entity_type_name(service_id) << dendl;
     ldout(cct, 30) << "_rotate_secret adding " << ceph_entity_type_name(service_id)
-	           << " id " << secret_id << " " << ek
-	           << dendl;
+               << " id " << secret_id << " " << ek
+               << dendl;
```
Comment (idryomov) [inline src/auth/cephx/CephxKeyServer.cc:214]: Unnecessary change.

File: src/auth/cephx/CephxKeyServer.cc [hunk 3]
```diff
@@ -210,9 +210,11 @@ int KeyServer::_rotate_secret(uint32_t service_id)
...
     added++;
+    if(auth)
```
Comment (idryomov) [inline src/auth/cephx/CephxKeyServer.cc:216]:
```suggestion
    if (auth) {
```

File: src/auth/cephx/CephxKeyServer.cc [hunk 4]
```diff
@@ -210,9 +210,11 @@ int KeyServer::_rotate_secret(uint32_t service_id)
...
+        auth = 0;
```
Comment (idryomov) [inline src/auth/cephx/CephxKeyServer.cc:217]:
```suggestion
        auth = 0;
      }  
```

File: src/mon/AuthMonitor.cc [hunk 5]
```diff
@@ -79,17 +79,41 @@ ostream& operator<<(ostream &out, const AuthMonitor &pm)
...
-  if (!mon.key_server.updated_rotating(rot_inc.rotating_bl, last_rotating_ver))
-    return false;
-  dout(10) << __func__ << " updated rotating" << dendl;
-  push_cephx_inc(rot_inc);
-  return true;
+  bool ret = false;
+  int  auth = get_auth_done();
+  if (check != CHECK_AUTH_DONE)
+    auth = NO_AUTH_PROPOSE;
+  if( auth == AUTH_PROPOSE_START){
```
Comment (idryomov) [inline src/mon/AuthMonitor.cc:90]:
```suggestion
  if (auth == AUTH_PROPOSE_START) {
```

File: src/mon/AuthMonitor.cc [hunk 6]
```diff
@@ -79,17 +79,41 @@ ostream& operator<<(ostream &out, const AuthMonitor &pm)
...
+    auth_propose_retry++;
+    dout(10) << __func__ << " updated rotating retry: " <<auth_propose_retry<< dendl;
```
Comment (idryomov) [inline src/mon/AuthMonitor.cc:92]:
```suggestion
    dout(10) << __func__ << " updated rotating retry: " << auth_propose_retry << dendl;
```

File: src/mon/AuthMonitor.cc [hunk 7]
```diff
@@ -79,17 +79,41 @@ ostream& operator<<(ostream &out, const AuthMonitor &pm)
...
+      auth_propose_retry = 0;
+      set_auth_done(NO_AUTH_PROPOSE);
+    }
+  }else if(auth == AUTH_PROPOSE_DONE){
+    dout(10) << __func__ << " updated rotating succeed" << dendl;
+    auth_propose_retry = 0;
+    auth = NO_AUTH_PROPOSE;
+    set_auth_done(NO_AUTH_PROPOSE);
+  }
+  if (!mon.key_server.updated_rotating(rot_inc.rotating_bl, last_rotating_ver,auth))
```
Comment (idryomov) [inline src/mon/AuthMonitor.cc:104]:
```suggestion
  if (!mon.key_server.updated_rotating(rot_inc.rotating_bl, last_rotating_ver, auth)) {
```

File: src/mon/AuthMonitor.h [hunk 8]
```diff
@@ -29,9 +29,12 @@ class KeyRing;
 class Monitor;
 
 #define MIN_GLOBAL_ID 0x1000
+#define MAX_AUTH_RETRY 0X3
+#define CHECK_AUTH_DONE 0X1
```
Comment (idryomov) [inline src/mon/AuthMonitor.h:33]: Just `1` and `3`, no need to use hexadecimal here.

**Review by idryomov — COMMENTED** [review]
There are multiple coding style issues, mostly related to missing spaces and braces.  I've highlighted a few instances but all of them would need to be fixed.

More importantly, I don't fully understand the problem you are trying to address yet.  I've asked for additional logs in the tracker ticket, let's continue discussing it there for now.

**Comment by swg1991** [comment]
@jecluis ping ? Do you have free time to  discuss it with us?

**Review by jecluis — CHANGES_REQUESTED** [review]
I don't particularly like the idea of changing `PaxosService` to accommodate a specific service's needs, especially when this seems like something that's not fully fleshed out. Following Ilya's statement, I'm also not entirely sure what it is you are trying to fix here?

**Pull request closed without merging**